### PR TITLE
Hyena Operator

### PR DESCRIPTION
**Pull request by guyjacob**
# What does this PR do ?

Add Hyena operator (https://arxiv.org/abs/2302.10866)

**Collection**: NLP

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/nlp/language_modeling/conf/megatron_gpt_config_hyena.yaml [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+  overlap_p2p_comm: False # Overlap p2p communication with computes. This argument is valid only when `virtual_pipeline_model_parallel_size` is larger than 1
+  batch_p2p_comm: True # Batch consecutive inter-peer send/recv operations. This argument is valid only when `virtual_pipeline_model_parallel_size` is larger than 1
+  seq_len_interpolation_factor: null # RoPE Interpolation factor for sequence length. This is used to build long-context models with RoPE ex: https://arxiv.org/abs/2306.15595.
+  num_query_groups: null # Number of query groups for group query attention. If None, normal attention is used.
+
+  name: te_gpt_hyena # key for selecting the correct ModuleSpec
+
+  hyena:
+    # HyenaOperator parameters
+    l_max: ${model.encoder_seq_length}
```
Comment (akoumpa) [inline examples/nlp/language_modeling/conf/megatron_gpt_config_hyena.yaml:101]: Hi, do you think it would be possible to find a more descriptive name for people less familiar with Hyena? I'm thinking `l_max` and `w` use a different name, thanks.

File: examples/nlp/language_modeling/conf/megatron_gpt_config_hyena.yaml [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+  # this feature removes most of activation checkpoints at the last pipeline stage, which is the critical execution path.
+
+  ## Sequence Parallelism
+  # Makes tensor parallelism more memory efficient for LLMs (20B+) by parallelizing layer norms and dropout sequentially
+  # See Reducing Activation Recomputation in Large Transformer Models: https://arxiv.org/abs/2205.05198 for more details.
+  sequence_parallel: False
+
+  ## Transformer Engine
+  transformer_engine: True
+  fp8: False # enables fp8 in TransformerLayer forward
```
Comment (akoumpa) [inline examples/nlp/language_modeling/conf/megatron_gpt_config_hyena.yaml:199]: is fp8 supported?

File: nemo/collections/nlp/modules/common/hyena/fftconv_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            dv = dv.to(dtype=v.dtype)  # We do atomicAdd in fp32 so might need to convert to fp16
+        return (
+            du,
+            dk,
+            dD,
+            None,
+            None,
+            None,
+            None,
+            dv if v is not None else None,
```
Comment (akoumpa) [inline nemo/collections/nlp/modules/common/hyena/fftconv_wrapper.py:105]: Hmm, would anything break if returned just `dv` here & similarly for `dq`?

File: nemo/collections/nlp/modules/common/hyena/fftconv_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
+import math
+
+import torch
+from einops import rearrange
+from fftconv import fftconv_bwd, fftconv_fwd
```
Comment (akoumpa) [inline nemo/collections/nlp/modules/common/hyena/fftconv_wrapper.py:5]: I wonder if this should be wrapped in try/catch exception block for people who don't have fftconv installed?

File: nemo/collections/nlp/modules/common/hyena/hyena.py [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+            order,
+            filter_order,
+            bias=bias,
+            filter_cls=filter_cls,
+            filter_submodules=filter_submodules,
+            **filter_kwargs,
+        )
+
+        if fftconv_type is None:
+            if l_max <= 8192 and HAVE_FFTCONV:
```
Comment (akoumpa) [inline nemo/collections/nlp/modules/common/hyena/hyena.py:157]: should we check if user uses l_max > 8k & fftconv_type = flash?

File: nemo/collections/nlp/modules/common/hyena/hyena.py [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+        bias = bias.to(dtype=torch.float32)
+        return safari_fftconv_fn(x, k, bias, gelu=False)
+
+    def _flash_fft(self, x, k, bias):
+        x = x.contiguous()
+        y = self.flashfftconv(x, k) + x * bias.unsqueeze(dim=1)
+        return y
+
+    def forward(self, x, k, recurrence_idx):
+        bias = self.bias if self.use_bias else 0 * self.bias
```
Comment (akoumpa) [inline nemo/collections/nlp/modules/common/hyena/hyena.py:186]: is there any way to avoid the 0 * self.bias? I wonder if changing fftconv_fn to accept bias=None would be possible here?

**Review by akoumpa — DISMISSED** [review]
LGTM, just a few minor comments, if you have to address these that would be great, thanks!

File: examples/nlp/language_modeling/conf/megatron_gpt_config_hyena.yaml (hunk 1 above)
Comment (guyjacob) [reply examples/nlp/language_modeling/conf/megatron_gpt_config_hyena.yaml]: Done (renamed and added more comments in code + yaml)

File: examples/nlp/language_modeling/conf/megatron_gpt_config_hyena.yaml (hunk 2 above)
Comment (guyjacob) [reply examples/nlp/language_modeling/conf/megatron_gpt_config_hyena.yaml:201]: The external FFT conv libraries don't support FP8

File: nemo/collections/nlp/modules/common/hyena/fftconv_wrapper.py (hunk 3 above)
Comment (guyjacob) [reply nemo/collections/nlp/modules/common/hyena/fftconv_wrapper.py]: Nothing would break - fixed. Thanks

File: nemo/collections/nlp/modules/common/hyena/hyena.py (hunk 5 above)
Comment (guyjacob) [reply nemo/collections/nlp/modules/common/hyena/hyena.py]: Done

File: nemo/collections/nlp/modules/common/hyena/hyena.py (hunk 6 above)
Comment (guyjacob) [reply nemo/collections/nlp/modules/common/hyena/hyena.py]: safari-fftconv expects a valid tensor. But definitely no reason to do this on every forward pass, moved it to `__init__`, creating a "vanilla" zero tensor (that is - not a `nn.Parameter`).

File: examples/nlp/language_modeling/conf/megatron_gpt_config_hyena.yaml (hunk 2 above)
Comment (akoumpa) [reply examples/nlp/language_modeling/conf/megatron_gpt_config_hyena.yaml:201]: ok, in that case, would adding an assert statement to throw an exception if it's enabled make sense?

**Review by akoumpa — DISMISSED** [review]

**Review by akoumpa — APPROVED** [review]
